File: Windows/programmers/Microwire.cs
using System;

namespace FlashcatUSB {
    public class Microwire_Programmer : FlashcatUSB.MemoryDeviceUSB {
        private FlashcatUSB.USB.FCUSB_DEVICE FCUSB;
        private FlashcatUSB.FlashMemory.MICROWIRE MICROWIRE_DEVIE;

        public event PrintConsoleEventHandler PrintConsole;

        public delegate void PrintConsoleEventHandler(string message);

        public event SetProgressEventHandler SetProgress;

        public delegate void SetProgressEventHandler(int percent);

        public Microwire_Programmer(FlashcatUSB.USB.FCUSB_DEVICE parent_if) {
            FCUSB = parent_if;
        }

        public bool DeviceInit() {
            var s93_devices = FlashcatUSB.MainApp.FlashDatabase.GetFlashDevices(FlashcatUSB.FlashMemory.MemoryType.SERIAL_MICROWIRE);
            MICROWIRE_DEVIE = null;
            uint org_mode = (uint)FlashcatUSB.MainApp.MySettings.S93_DEVICE_ORG; // 0=8-bit,1=16-bit
            if (!FlashcatUSB.MainApp.MySettings.S93_DEVICE.Equals("")) {
                foreach (var device in s93_devices) {
                    if (device.NAME.ToUpper().Equals(FlashcatUSB.MainApp.MySettings.S93_DEVICE.ToUpper())) {
                        MICROWIRE_DEVIE = (FlashcatUSB.FlashMemory.MICROWIRE)device;
                        break;
                    }
                }
            }

            if (MICROWIRE_DEVIE is null) {
                PrintConsole?.Invoke("No Microwire device selected");
                FlashcatUSB.MainApp.SetStatus("No Microwire device selected");
                return false;
            }

            uint addr_bits = 0U;
            string org_str;
            if (org_mode == 0L) { // 8-bit
                org_str = "X8";
                addr_bits = (uint)MICROWIRE_DEVIE.X8_ADDRSIZE;
            } else { // 16-bit mode
                org_str = "X16";
                addr_bits = (uint)MICROWIRE_DEVIE.X16_ADDRSIZE;
            }

            PrintConsole?.Invoke("Microwire device: " + DeviceName + " (" + MICROWIRE_DEVIE.FLASH_SIZE + " bytes) " + org_str + " mode");
            uint setup_data = addr_bits << 8 | org_mode;
            bool result = FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.S93_INIT, null, setup_data);
            return result;
        }

        public string DeviceName {
            get {
                return MICROWIRE_DEVIE.NAME;
            }
        }

        public long DeviceSize {
            get {
                return MICROWIRE_DEVIE.FLASH_SIZE;
            }
        }

        public uint SectorSize(uint sector, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return (uint)DeviceSize;
        }

        public byte[] ReadData(long flash_offset, long data_count, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            try {
                var setup_data = new byte[8];
                bool result;
                setup_data[0] = (byte)(data_count >> 24 & 255L);
                setup_data[1] = (byte)(data_count >> 16 & 255L);
                setup_data[2] = (byte)(data_count >> 8 & 255L);
                setup_data[3] = (byte)(data_count & 255L);
                setup_data[4] = (byte)(flash_offset >> 24 & 255L);
                setup_data[5] = (byte)(flash_offset >> 16 & 255L);
                setup_data[6] = (byte)(flash_offset >> 8 & 255L);
                setup_data[7] = (byte)(flash_offset & 255L);
                var data_out = new byte[(int)(data_count - 1L + 1)];
                result = FCUSB.USB_SETUP_BULKIN(FlashcatUSB.USB.USBREQ.S93_READEEPROM, setup_data, ref data_out, 0U);
                if (!result)
                    return null;
                return data_out;
            } catch (Exception ex) {
            }

            return null;
        }

        public bool WriteData(long flash_offset, byte[] data_to_write, FlashcatUSB.WriteParameters Params = null) {
            try {
                uint data_count = (uint)data_to_write.Length;
                var setup_data = new byte[8];
                bool result;
                setup_data[0] = (byte)(data_count >> 24 & 255L);
                setup_data[1] = (byte)(data_count >> 16 & 255L);
                setup_data[2] = (byte)(data_count >> 8 & 255L);
                setup_data[3] = (byte)(data_count & 255L);
                setup_data[4] = (byte)(flash_offset >> 24 & 255L);
                setup_data[5] = (byte)(flash_offset >> 16 & 255L);
                setup_data[6] = (byte)(flash_offset >> 8 & 255L);
                setup_data[7] = (byte)(flash_offset & 255L);
                result = FCUSB.USB_SETUP_BULKOUT(FlashcatUSB.USB.USBREQ.S93_WRITEEEPROM, setup_data, data_to_write, data_count);
                FlashcatUSB.Utilities.Main.Sleep(100);
                FCUSB.USB_WaitForComplete();
                if (result)
                    ReadData(0L, 16L); // Some devices need us to read a page of data
                return result;
            } catch (Exception ex) {
            }

            return false;
        }

        public bool EraseDevice() {
            bool result = FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.S93_ERASE);
            return result;
        }

        public void WaitUntilReady() {
            FlashcatUSB.Utilities.Main.Sleep(10);
        }

        public long SectorFind(uint SectorIndex, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return 0L;
        }

        public bool SectorErase(uint SectorIndex, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return true;
        }

        public uint SectorCount() {
            return 1U;
        }

        public bool SectorWrite(uint SectorIndex, byte[] data, FlashcatUSB.WriteParameters Params = null) {
            return WriteData(0L, data);
        }
    }
}

File: Windows/programmers/Microwire.vb
Imports FlashcatUSB.FlashMemory
Public Class Microwire_Programmer : Implements MemoryDeviceUSB
    Private FCUSB As USB.FCUSB_DEVICE

    Private MICROWIRE_DEVIE As MICROWIRE

    Public Event PrintConsole(message As String) Implements MemoryDeviceUSB.PrintConsole
    Public Event SetProgress(percent As Integer) Implements MemoryDeviceUSB.SetProgress

    Sub New(parent_if As USB.FCUSB_DEVICE)
        FCUSB = parent_if
    End Sub

    Public Function DeviceInit() As Boolean Implements MemoryDeviceUSB.DeviceInit
        Dim s93_devices() As Device = FlashDatabase.GetFlashDevices(MemoryType.SERIAL_MICROWIRE)
        Me.MICROWIRE_DEVIE = Nothing
        Dim org_mode As UInt32 = MySettings.S93_DEVICE_ORG '0=8-bit,1=16-bit
        If Not MySettings.S93_DEVICE.Equals("") Then
            For Each mem_device In s93_devices
                If mem_device.NAME.ToUpper.Equals(MySettings.S93_DEVICE.ToUpper) Then
                    Me.MICROWIRE_DEVIE = mem_device
                    Exit For
                End If
            Next
        End If
        If Me.MICROWIRE_DEVIE Is Nothing Then
            RaiseEvent PrintConsole("No Microwire device selected")
            Return False
        End If
        Dim addr_bits As UInt32 = 0
        Dim org_str As String
        If org_mode = 0 Then '8-bit
            org_str = "X8"
            addr_bits = Me.MICROWIRE_DEVIE.X8_ADDRSIZE
        Else '16-bit mode
            org_str = "X16"
            addr_bits = Me.MICROWIRE_DEVIE.X16_ADDRSIZE
        End If
        RaiseEvent PrintConsole("Microwire device: " & Me.DeviceName & " (" & Me.MICROWIRE_DEVIE.FLASH_SIZE & " bytes) " & org_str & " mode")
        Dim setup_data As UInt32 = (addr_bits << 8) Or (org_mode)
        Dim result As Boolean = FCUSB.USB_CONTROL_MSG_OUT(USB.USBREQ.S93_INIT, Nothing, setup_data)
        Return result
    End Function

    Public ReadOnly Property GetDevice As Device Implements MemoryDeviceUSB.GetDevice
        Get
            Return Me.MICROWIRE_DEVIE
        End Get
    End Property

    Public ReadOnly Property DeviceName As String Implements MemoryDeviceUSB.DeviceName
        Get
            Return Me.MICROWIRE_DEVIE.NAME
        End Get
    End Property

    Public ReadOnly Property DeviceSize As Long Implements MemoryDeviceUSB.DeviceSize
        Get
            Return Me.MICROWIRE_DEVIE.FLASH_SIZE
        End Get
    End Property

    Public Function SectorSize(sector As UInteger) As UInt32 Implements MemoryDeviceUSB.SectorSize
        Return Me.DeviceSize
    End Function

    Public Function ReadData(flash_offset As Long, data_count As Long) As Byte() Implements MemoryDeviceUSB.ReadData
        Try
            Dim setup_data(7) As Byte
            Dim result As Boolean
            setup_data(0) = ((data_count >> 24) And 255)
            setup_data(1) = ((data_count >> 16) And 255)
            setup_data(2) = ((data_count >> 8) And 255)
            setup_data(3) = (data_count And 255)
            setup_data(4) = ((flash_offset >> 24) And 255)
            setup_data(5) = ((flash_offset >> 16) And 255)
            setup_data(6) = ((flash_offset >> 8) And 255)
            setup_data(7) = (flash_offset And 255)
            Dim data_out(data_count - 1) As Byte
            result = FCUSB.USB_SETUP_BULKIN(USB.USBREQ.S93_READEEPROM, setup_data, data_out, 0)
            If Not result Then Return Nothing
            Return data_out
        Catch ex As Exception
        End Try
        Return Nothing
    End Function

    Public Function WriteData(flash_offset As Long, data_to_write() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.WriteData
        Try
            Dim data_count As UInt32 = data_to_write.Length
            Dim setup_data(7) As Byte
            Dim result As Boolean
            setup_data(0) = ((data_count >> 24) And 255)
            setup_data(1) = ((data_count >> 16) And 255)
            setup_data(2) = ((data_count >> 8) And 255)
            setup_data(3) = (data_count And 255)
            setup_data(4) = ((flash_offset >> 24) And 255)
            setup_data(5) = ((flash_offset >> 16) And 255)
            setup_data(6) = ((flash_offset >> 8) And 255)
            setup_data(7) = (flash_offset And 255)
            result = FCUSB.USB_SETUP_BULKOUT(USB.USBREQ.S93_WRITEEEPROM, setup_data, data_to_write, data_count)
            Utilities.Sleep(100)
            FCUSB.USB_WaitForComplete()
            If result Then ReadData(0, 16) 'Some devices need us to read a page of data
            Return result
        Catch ex As Exception
        End Try
        Return False
    End Function

    Public Function EraseDevice() As Boolean Implements MemoryDeviceUSB.EraseDevice
        Dim result As Boolean = FCUSB.USB_CONTROL_MSG_OUT(USB.USBREQ.S93_ERASE)
        Return result
    End Function

    Public Sub WaitUntilReady() Implements MemoryDeviceUSB.WaitUntilReady
        Utilities.Sleep(10)
    End Sub

    Public Function SectorFind(SectorIndex As UInt32) As Long Implements MemoryDeviceUSB.SectorFind
        Return 0
    End Function

    Public Function SectorErase(SectorIndex As UInteger) As Boolean Implements MemoryDeviceUSB.SectorErase
        Return True
    End Function

    Public Function SectorCount() As UInteger Implements MemoryDeviceUSB.SectorCount
        Return 1
    End Function

    Public Function SectorWrite(SectorIndex As UInteger, data() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.SectorWrite
        Return WriteData(0, data)
    End Function


End Class
